src/ucp/dt/dt_struct.c

```diff
@@ -0,0 +1,365 @@
+        }
+    }
+
+    /* Sanity check:
+     * Structured datatype only supports UCP_DATATYPE_CONTIG and
+     * UCP_DATATYPE_STRUCT as sub-datatypes
+     */
+
+    ret = ucs_posix_memalign((void **)&dt,
+                             ucs_max(sizeof(void *), UCS_BIT(UCP_DATATYPE_SHIFT)),
+                             sizeof(*dt), "struct_dt");
+    if (ret != 0) {
+        return UCS_ERR_NO_MEMORY;
+    }
+
+    ret = ucs_posix_memalign((void **)&dt->desc, sizeof(*dt->desc),
+                             sizeof(*dt->desc) * dt->desc_count,
+                             "ucp_dt_struct_t");
+    if (ret != 0) {
+        ucs_free(dt);
+        return UCS_ERR_NO_MEMORY;
+    }
+    memcpy(dt->desc, desc_ptr, sizeof(*desc_ptr) * desc_count);
+    dt->desc_count = desc_count;
+    dt->rep_count = rep_count;
+    _set_length(dt);
+    _set_depth(dt);
+    *datatype_p = ((uintptr_t)dt) | UCP_DATATYPE_STRUCT;
+    return UCS_OK;
+}
+
+void ucp_dt_destroy_struct(ucp_datatype_t datatype_p)
+{
+    ucp_dt_struct_t *dt;
+    dt = ucp_dt_struct(datatype_p);
+    ucs_free(dt->desc);
+    ucs_free(dt);
+}
+
+void ucp_dt_struct_gather(void *dest, const void *src, ucp_datatype_t dt,
+                          size_t length, size_t offset, void *state)
+{
+    size_t processed_len;
+    ucp_dt_struct_t *s = ucp_dt_struct(dt);
+    /* TODO: enable using "state" to make it more efficient.
+     * Right now it always performs the "seek" operation which is
+     * inefficient
+     */
+     processed_len = _dte_pack(s, src, dest, offset, length);
+
+     /* We assume that the sane length was provided */
+     ucs_assert(processed_len == length);
+}
+
+size_t ucp_dt_struct_scatter(void *dst, ucp_datatype_t dt,
+                             const void *src, size_t length, size_t offset,
+                             void *state)
+{
+    size_t processed_len;
+    ucp_dt_struct_t *s = ucp_dt_struct(dt);
+    /* TODO: enable using "state" to make it more efficient.
+     * Right now it always performs the "seek" operation which is
+     * inefficient
+     */
+     processed_len = _dte_unpack(s, src, dst, offset, length);
+
+     /* We assume that the sane length was provided */
+     ucs_assert(processed_len == length);
+     return processed_len;
+}
+
+/* Dealing with UCT */
+
+inline static void _to_cache(ucp_dt_struct_t *s, void *ptr, uct_mem_h memh)
+{
+    khiter_t k;
+    int ret;
+    k = kh_put(dt_struct, &s->hash, (uint64_t)ptr, &ret);
+    /* TODO: check ret */
+    kh_value(&s->hash, k) = memh;
+}
+inline static int _in_cache(ucp_dt_struct_t *s, void *ptr)
+{
+    khiter_t k;
+    k = kh_get(dt_struct, &s->hash, (uint64_t)ptr);
+    return (k != kh_end(&s->hash));
+}
+
+static int _is_primitive_strided(ucp_dt_struct_t *s)
+{
+    size_t i;
+    /* Has only 2 levels of nesting */
+    if (!(s->depth = 2)){
+        return 0;
+    }
+
+    /* On the leaf level, all datatypes are contig */
+    for (i = 0; i < s->desc_count; i++) {
```

Comment:
so, we do not allow nesting striding?
worth to check and print error during dt creation